### PR TITLE
refactor(component): Rewrite `Sidebar` as a compound component

### PR DESCRIPTION
**Pull request by tulup-conner**
Refactors the `Sidebar` component into a compound component.

This is a dramatic API change. `Sidebar`s used to be constructed like so:

```js
<Sidebar
  collapsed={false}
  itemsGroups={[
    [
      {
        group: false,
        href: '#',
        icon: HiChartPie,
        title: 'Dashboard'
      },
      {
        group: false,
        href: '#',
        icon: HiViewBoards,
        label: 'Pro',
        title: 'Kanban'
      },
      {
        group: false,
        href: '#',
        icon: HiInbox,
        label: '3',
        title: 'Inbox'
      },
      {
        group: false,
        href: '#',
        icon: HiUser,
        title: 'Users'
      },
      {
        group: false,
        href: '#',
        icon: HiShoppingBag,
        title: 'Products'
      },
      {
        group: false,
        href: '#',
        icon: HiArrowSmRight,
        title: 'Sign In'
      },
      {
        group: false,
        href: '#',
        icon: HiTable,
        title: 'Sign Up'
      }
    ]
  ]}
/>
```

They now look like this:

```js
<Sidebar>
  <Sidebar.Items>
    <Sidebar.ItemGroup>
      <Sidebar.Item href="#" icon={HiChartPie}>
        Dashboard
      </Sidebar.Item>
      <Sidebar.Item href="#" icon={HiViewBoards}>
        Kanban
      </Sidebar.Item>
      <Sidebar.Item href="#" icon={HiInbox}>
        Inbox
      </Sidebar.Item>
      <Sidebar.Item href="#" icon={HiUser}>
        Users
      </Sidebar.Item>
      <Sidebar.Item href="#" icon={HiShoppingBag}>
        Products
      </Sidebar.Item>
      <Sidebar.Item href="#" icon={HiArrowSmRight}>
        Sign In
      </Sidebar.Item>
      <Sidebar.Item href="#" icon={HiTable}>
        Sign Up
      </Sidebar.Item>
    </Sidebar.ItemGroup>
  </Sidebar.Items>
</Sidebar>
```

This patch also introduces the ability to add a CTA and/or logo to the `Sidebar` like so:

```js
<Sidebar>
  <Sidebar.Logo img="logo192.png" href="#">
    Flowbite
  </Sidebar.Logo>
  <Sidebar.Items>...</Sidebar.Items>
</Sidebar>
```

```js
<Sidebar>
  <Sidebar.CTA color="purple">
    <Button>Hello</Button>      
  </Sidebar.CTA>
  <Sidebar.Items>...</Sidebar.Items>
</Sidebar>
```

### CODE REVIEW

**Comment by tulup-conner** [comment]
I need to rewrite the `root` implementation of Sidebar to use this new API, then it should be ready to review.

File: src/components/Sidebar/SidebarItem.tsx [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import classNames from 'classnames';
+import { ComponentProps, FC, PropsWithChildren } from 'react';
+import { Link, useLocation } from 'react-router-dom';
+import { SidebarProps } from '.';
+import { Badge, BadgeColor } from '../Badge';
+export interface SidebarItemProps extends PropsWithChildren<Pick<SidebarProps, 'collapsed'>> {
```
Comment (rluders) [inline src/components/Sidebar/SidebarItem.tsx:6]: one space here to split the imports :)

File: src/components/Sidebar/SidebarCTA.tsx [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import classNames from 'classnames';
+import { PropsWithChildren, FC } from 'react';
+import { Color } from '../Button';
+
+interface SidebarCTAProps extends PropsWithChildren<Record<string, unknown>> {
```
Comment (rluders) [inline src/components/Sidebar/SidebarCTA.tsx:5]: I guess that @bacali95 will say: use type instead interface :)

File: src/components/Sidebar/SidebarItemGroup.tsx [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import classNames from 'classnames';
+import { cloneElement, FC, ReactElement } from 'react';
+import { SidebarProps } from '.';
+import childrenAsArray from '../../helpers/childrenAsArray';
+
+interface SidebarItemGroupProps extends SidebarProps {
```
Comment (rluders) [inline src/components/Sidebar/SidebarItemGroup.tsx:6]: same thing here... use types.

File: src/components/Sidebar/SidebarLogo.tsx [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import { PropsWithChildren, FC } from 'react';
+
+interface SidebarLogoProps extends PropsWithChildren<Record<string, unknown>> {
```
Comment (rluders) [inline src/components/Sidebar/SidebarLogo.tsx:3]: again types. :)

File: src/components/Sidebar/index.tsx [hunk 5]
```diff
@@ -1,150 +1,39 @@
...
-
-  const ItemWrapper = ({ title, children }: { title: string; children: ReactNode }) =>
-    collapsed ? (
-      <Tooltip content={title} placement="right">
-        {children}
-      </Tooltip>
-    ) : (
-      <>{children}</>
-    );
+export interface SidebarProps extends PropsWithChildren<Record<string, unknown>> {
```
Comment (rluders) [inline src/components/Sidebar/index.tsx:11]: types

File: src/helpers/childrenAsArray.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import { ReactNode } from 'react';
+
+export default function childrenAsArray(children: ReactNode): ReactNode[] {
+  if (Array.isArray(children)) {
+    return children;
+  } else if (typeof children !== 'undefined') {
```
Comment (rluders) [inline src/helpers/childrenAsArray.ts:6]: I don't like `else-if` statement, and you don't need a simple `if` would do the job

File: src/helpers/childrenAsArray.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+    return [children];
+  } else {
```
Comment (rluders) [inline src/helpers/childrenAsArray.ts:8]: you don't need the `else`, 'cause you are already returning if it fits in any other flow. :)

**Review by rluders — CHANGES_REQUESTED** [review]
Wow! Great job. I liked all the changes that you made, the adopted approach gives much more flexibility to the component, and it looks like the other components as well.

I added a few comments to suggest some changes, mostly related to code style and one code improvement. :)

File: src/components/Sidebar/SidebarCTA.tsx (hunk 2 above)
Comment (bacali95) [reply src/components/Sidebar/SidebarCTA.tsx:5]: Yes I will say that indeed :slightly_smiling_face:

**Comment by tulup-conner** [comment]
@bacali95 Can you explain not using interface? It is recommended by Typescript https://www.typescriptlang.org/play#example/types-vs-interfaces so I am curious.

**Comment by bacali95** [comment]
> @bacali95 Can you explain not using interface? It is recommended by Typescript https://www.typescriptlang.org/play#example/types-vs-interfaces so I am curious.

Technically there is not that big of a difference, it is just a personal choice that I consider `types` are more readable and easier to work with, but like I said it is just a personal choice and this library is not my personal project anymore, so I will not stop anyone from using `interfaces` but then we need to use it over all the project to keep the code consistent :wink:

**Comment by rluders** [comment]
> > @bacali95 Can you explain not using interface? It is recommended by Typescript https://www.typescriptlang.org/play#example/types-vs-interfaces so I am curious.
> 
> Technically there is not that big of a difference, it is just a personal choice that I consider `types` are more readable and easier to work with, but like I said it is just a personal choice and this library is not my personal project anymore, so I will not stop anyone from using `interfaces` but then we need to use it over all the project to keep the code consistent wink

Well, if we have to choose... I would say +1 to use interface.

**Review by rluders — APPROVED** [review]
lgtm

**Pull request merged**